### PR TITLE
API-25775 Upgrade AWS SDK DynamoDB from v2 to v3 and upgrade unit tests

### PR DESCRIPTION
**Pull request by mwise-va**
Jira Story: 

https://vajira.max.gov/browse/API-25775

Title: 

API-25775 Upgrade AWS SDK DynamoDB from v2 to v3 and upgrade unit tests

Description:

This pull request upgrades the AWS SDK DynamoDB dependency from version 2 to version 3, which improves performance, better error handling, and a more intuitive API. Additionally I have upgraded 10 unit tests that use the upgraded AWS SDK to ensure the functionality of the upgraded library.

To upgrade the SDK, I made the following changes:

1. Updated the SDK dependency in the 'package.json' to aws-sdk@3.x.x'
2. Replaced all references to the v2 SDK with the equivalent code using the v3 SDK
3. Updated the syntax of the code to comply with the new API
4. Removed any deprecated functionality used in the previous version
5. Updated Dynamo client unit tests to comply with v3 code standards
6. Isolated Dynamo schema

Furthermore, I updated the Dynamo client unit tests to verify the correct behavior of the upgraded DynamoDB library. The unit tests ensures all Dynamo client operations complete successfully without any errors.

Please let me know if you have any suggestions or feedback.

References:

- AWS SDK v3 DynamoDBDocumentClient.   https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide/dynamodb-example-dynamodb-utilities.html
- AWS SDK v3 Client Mock.   https://www.npmjs.com/package/aws-sdk-client-mock

Thanks,

Mitch

### CODE REVIEW

File: src/dynamo_client.js [hunk 1]
```diff
@@ -32,23 +58,20 @@ class DynamoClient {
...
-        } else {
-          resolve(data);
-        }
-      });
-    });
+    try {
+      const data = await this.ddbDocClientV3.send(new PutCommand(params));
+      return data;
+    } catch (err) {
+      console.log("Error", err);
```
Comment (crolarlibertyva) [inline src/dynamo_client.js:73]: This should be a throw. Remove the "console.log"

File: src/dynamo_client.js [hunk 2]
```diff
@@ -58,20 +81,17 @@ class DynamoClient {
...
-        } else {
-          resolve(data);
-        }
-      });
-    });
+    try {
+      const data = await this.ddbDocClientV3.send(new ScanCommand(params));
+      return data;
+    } catch (err) {
+      console.log("Error", err);
```
Comment (crolarlibertyva) [inline src/dynamo_client.js:93]: This should be a throw. Remove the "console.log"

File: src/dynamo_client.js [hunk 3]
```diff
@@ -82,23 +102,20 @@ class DynamoClient {
...
-        } else {
-          resolve(data);
-        }
-      });
-    });
+    try {
+      const data = await this.ddbDocClientV3.send(new GetCommand(params));
+      return data;
+    } catch (err) {
+      console.log("Error", err);
```
Comment (crolarlibertyva) [inline src/dynamo_client.js:117]: This should be a throw. Remove the "console.log"

File: src/dynamo_client.js [hunk 4]
```diff
@@ -131,15 +148,12 @@ class DynamoClient {
...
-        } else {
-          resolve(data);
-        }
-      });
-    });
+    try {
+      const data = await this.ddbDocClientV3.send(new QueryCommand(params));
+      return data;
+    } catch (err) {
+      console.log("Error", err);
```
Comment (crolarlibertyva) [inline src/dynamo_client.js:155]: This should be a throw. Remove the "console.log"

File: src/dynamo_client.js [hunk 5]
```diff
@@ -166,15 +180,12 @@ class DynamoClient {
...
-        } else {
-          resolve(data);
-        }
-      });
-    });
+    try {
+      const data = await this.ddbDocClientV3.send(new UpdateCommand(params));
+      return data;
+    } catch (err) {
+      console.log("Error", err);
```
Comment (crolarlibertyva) [inline src/dynamo_client.js:187]: This should be a throw. Remove the "console.log"

**Review by crolarlibertyva — DISMISSED** [review]
Beside the catch section, some tests are failing.

**Comment by dbrown-va** [comment]
For an upgrade like this, it would be good to link the corresponding AWS documentation for the differences with V2 and V3 of the SDK as well as a short description of the expected impact.

Also include a direct link to the JIRA issue in the description.

File: src/dynamo_client.js (hunk 1 above)
Comment (mwise-va) [reply src/dynamo_client.js]: I switched the console.log("Error", err); with the following:

throw { error: err };

File: src/dynamo_client.js (hunk 2 above)
Comment (mwise-va) [reply src/dynamo_client.js]: I switched the console.log("Error", err); with the following:

throw { error: err };

File: src/dynamo_client.js (hunk 3 above)
Comment (mwise-va) [reply src/dynamo_client.js]: I switched the console.log("Error", err); with the following:

throw { error: err };

File: src/dynamo_client.js (hunk 4 above)
Comment (mwise-va) [reply src/dynamo_client.js]: I switched the console.log("Error", err); with the following:

throw { error: err };

File: src/dynamo_client.js (hunk 5 above)
Comment (mwise-va) [reply src/dynamo_client.js]: I switched the console.log("Error", err); with the following:

throw { error: err };

File: .gitignore [hunk 6]
```diff
@@ -140,4 +140,13 @@ tests/coverage
...
-.dccache
\ No newline at end of file
+.dccache
+
+#localdb
+localdb/package-lock.json
+
+#package-lock.json exclude
+"exclude": {
+    "files": ./package-lock.json",
```
Comment (crolarlibertyva) [inline .gitignore:150]: Any reason for this exclude?

**Review by crolarlibertyva — APPROVED** [review]
Would like a followup ticket that alters the unit tests that mock the dynamoclient to mimic its behavior more closely, ie. the new chaned promise structure vs returning a promise.

File: .gitignore (hunk 6 above)
Comment (mwise-va) [reply .gitignore]: Was having some issues with committing my package-lock.json file. I have fixed these changes and removed the exclude.

File: localdb/package.json [hunk 7]
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "oauth-proxy",
+  "version": "1.0.0",
+  "description": "Proxies requests to Okta to conform to FHIR API standard",
```
Comment (dbrown-va) [inline localdb/package.json:4]: For this localdb specific package let's update the name & description to align with the local only db functionality

**Comment by dbrown-va** [comment]
More out of curiosity after reading about the V3 upgrade, do you have image sizes before & after the upgrade to show the reduction for getting only the required AWS SDK packages?

File: src/dynamo_client.js [hunk 8]
```diff
@@ -32,23 +58,20 @@ class DynamoClient {
    * @returns {Promise()} this function returns the params payload to dynamo contained within the Promise constructor
```
Comment (dbrown-va) [inline src/dynamo_client.js:58]: This method is no longer returning a promise of the resulting payload.

The javadoc should be updated to reflect that change.

Comment (crolarlibertyva) [reply src/dynamo_client.js:58]: Good catch. I missed that.

**Comment by crolarlibertyva** [comment]
> More out of curiosity after reading about the V3 upgrade, do you have image sizes before & after the upgrade to show the reduction for getting only the required AWS SDK packages?

Local exported docker image compare
```
-rw-r--r--@ 1 larrycrochetjr  staff   1.6G May  1 21:00 /Users/larrycrochetjr/Downloads/oauth-proxy-aws-sdk-2.tar
-rw-r--r--  1 larrycrochetjr  staff   1.5G May  1 21:03 /Users/larrycrochetjr/Downloads/oauth-proxy-aws-sdk-3.tar
```
Looks to be about 1G.

File: src/dynamo_client.js (hunk 8 above)
Comment (mwise-va) [reply src/dynamo_client.js:58]: I updated the comments for each function in dynamo_client.js. Thanks for catching this.

**Review by crolarlibertyva — APPROVED** [review]

**Pull request merged**